### ISSUE TITLE
I2I: vendor specific batching plugin

### ISSUE THREAD

**Issue by lannka** [ampproject/amphtml#12432]
This is a follow up of the "batching support" proposal (#11241).
 
We will introduce “batching plugins” to amp-analytics to perform vendor specific aggregation logic on the data that's going to be sent. This can be helpful for vendors to do their own URL formatting, events dedup, etc.

A “batching plugin” is simply a callback function that is checked into the amp-analytics code base, which will be called right before a batch ping is going to be sent to compute the final destination URL.

It is configured like the following:

in `vendors.js`
```
requests: {
  heartbeat: {
    "baseUrl": "https://fake.com/batch-analytics?RANDOM",
    "maxDelay": 5,
    "batchPlugin": "some-vendor"
  }
}

triggers: {
  heartbeat: {
    on: "timer"
    request: "heartbeat",
    ...
    extraUrlParams: {
       param_a: 123,
       param_b: “xyz” 
    }
  }
}
```

in batch-plugins.js (newly proposed file for all the batching plugins)
```
"some-vendor": function(params) {
  // ...
  return url;
}
```
Where `params` is an array of BatchSegmentData defined as:
```
BatchSegmentData: {
  baseUrl: “https://fake.com/batch-analytics?0.1234”,
  trigger: “heartbeat”,
  timestamp: 149372837482734827,
  extraUrlParams: {
    param_a: 123,
    param_b: “xyz” 
  }
}
```

/cc @cramforce @pomeroyr

**Comment by zhouyx** [comment]
Before we go ahead and support batching plugins, I have 2 main questions. 
@cramforce @lannka 

**1. Who are qualified to add plugin to our code base?** 
- Big vendors / One vendor with multi plugins / Site owner ?
- What's our bar here? 

**2. What's the standard for batching plugins?**
We support batching plugins because some vendors need to perform their specific aggregation logic. 
- Logic besides aggregation?
For example: One could apply `toUpperCase` to every param in batching plugin functions. Which can be achieved by analytics filter functions. (consider doing that in batching plugin functions is much easier for vendors). 
- Plugin size limit? 
What's the size limit we are applying. And can we lazily load batching plugins.

**Comment by lannka** [comment]
My thoughts. @cramforce your call

> 1. Who are qualified to add plugin to our code base?

Plugin has to be designed for one or more vendors. So we don't accept any plugin designed only for a specific publisher.

We might enforce this by not allowing `batchPlugin` configured in pub config or remote config (only vendors.js). 

> 2. What's the standard for batching plugins?

Every new plugin introduced has to be proposed and reviewed by AMP team. We avoid a plugin as much as possible if it can be addressed with a more generic approach.

**Comment by cramforce** [comment]
We can do this on a case by case basis for now. Generally: Same rules as for any new extension.

**Comment by rudygalfi** [comment]
Should this be closed? If no, can you list remaining tasks?

**Comment by lannka** [comment]
We can close, since the doc is now merged #13065
